File: web/src/plugin-api/share.ts
/**
 * Shared artifacts（通用插件能力）—— 插件把自己的内容发布成「自带凭据的分享链接」。
 *
 * 宿主对内容零感知：`contentType` 由产出插件命名、`payload` 对宿主不透明，
 * 只有该插件自己注册的 share renderer 能解释它。`token` 即凭据（256 位高熵
 * 随机），任何人拿到链接即可读取这一份快照，**无需登录** —— 这正是「分享给朋友」
 * 的语义。
 *
 * 与 messageRenderer 同一模式：能力由插件**声明**，宿主只做通用派发，核心代码
 * 不出现任何具体内容类型（不感知 genui / 任何其他插件）。
 */
import type { ReactNode } from 'react'

/** 一份不可变的分享快照。 */
export interface SharedArtifact {
  token: string
  /** 产出该 artifact 的插件 id —— 公开分享页据此加载对应插件取得渲染器。 */
  pluginId: string
  /** 产出插件命名，如 'xbot.genui/tsx'。 */
  contentType: string
  /** 对宿主不透明；只有对应 share renderer 能解释。 */
  payload: string
  title: string
  createdAt: string
  /** RFC3339，或 '' = 永不过期。 */
  expiresAt: string
}

/** create() 的返回：artifact + 可直接分享的站内路径。 */
export interface ShareLink extends SharedArtifact {
  /** 站内路径，如 `/s/<token>`（拼上站点域名即可分享）。 */
  path: string
}

/** 公开渲染器：宿主在 /s/:token 页按 contentType 派发到它。 */
export interface ShareRendererContribution {
  kind: 'shareRenderer'
  id: string
  /** 负责的 artifact 类型，与 create() 传入的 contentType 一致。 */
  contentType: string
  /** 渲染一份 artifact。返回 null 表示「不是我负责的」，宿主继续 fallback。 */
  render: (artifact: SharedArtifact) => ReactNode | null
}

export interface CreateShareInput {
  contentType: string
  payload: string
  title?: string
  /** <= 0 或省略 = 永不过期。 */
  expiresInDays?: number
}

export interface ShareAPI {
  /** 发布一份快照，返回可直接分享的链接。 */
  create(input: CreateShareInput): Promise<ShareLink>
  /** 列出本插件（当前会话）创建的分享。 */
  list(): Promise<ShareLink[]>
  /** 撤销一份分享（立即失效）。 */
  revoke(token: string): Promise<void>
  /** 注册公开渲染器；宿主在分享页据 contentType 派发。返回退订函数。 */
  registerRenderer(decl: ShareRendererContribution): () => void
}

File: protocol/share.go
package protocol

// SharedArtifact is the wire contract for the generic shared-artifact
// capability: a plugin may publish an immutable snapshot of its own content and
// hand out a link that grants read access without any session.
//
// The host is content-agnostic by design — ContentType is named by the
// producing plugin (e.g. "xbot.genui/tsx") and Payload is opaque to the host;
// only that plugin's own share renderer can interpret it. Token is the
// credential: a high-entropy random string that resolves to this one artifact.
type SharedArtifact struct {
	Token string `json:"token"`
	// PluginID is the producing plugin — the public share page loads exactly
	// that plugin to obtain the renderer for this artifact.
	PluginID    string `json:"plugin_id"`
	ContentType string `json:"content_type"`
	Payload     string `json:"payload"`
	Title       string `json:"title"`
	CreatedAt   string `json:"created_at"`
	CreatedBy   int64  `json:"created_by"`
	// ExpiresAt is RFC3339, or "" for no expiry.
	ExpiresAt string `json:"expires_at,omitempty"`
	// RevokedAt is RFC3339 once revoked, otherwise "".
	RevokedAt string `json:"revoked_at,omitempty"`
}

// ShareCreateRequest is the request body of POST /api/share/create.
type ShareCreateRequest struct {
	// PluginID is filled by the plugin runtime (ctx.share knows it from
	// ctx.meta) — plugins never pass it themselves.
	PluginID string `json:"plugin_id"`
	// ContentType must be a value the calling plugin owns; the host only
	// validates that it is non-empty and bounded in length.
	ContentType string `json:"content_type"`
	// Payload is the opaque snapshot handed back to the plugin's renderer.
	Payload string `json:"payload"`
	// Title is display metadata for share listings.
	Title string `json:"title,omitempty"`
	// ExpiresInDays <= 0 means the link does not expire.
	ExpiresInDays int `json:"expires_in_days,omitempty"`
}
